### DISCUSSION TITLE
Wiring up form-level error messages

### DISCUSSION THREAD

**Discussion by delano** [vueform/vueform#28] [Questions] [answered]
First off, I'm really enjoying working with Vueform. There's a lot of nuanced functionality without feeling like it's over-engineered. Very cool 🙆 

I have a question about handling form-level errors -- specifically the kind that involves a round trip to the API and back. The field-level validation and error messages are working just super well out of the box. I think I'm missing something at the form-level though. 

We have a multi-step form that looks like this: 

```vue
<Vueform endpoint="/api/accounts/register/">
  <template #empty>
    <FormErrors v-if="hasErrors" />

    <FormSteps>
      ...
    </FormSteps>

    <FormElements>
      ...
    </FormElements>

    <FormStepsControls :labels="false">
      <template #previous><div class="flex"><ArrowLeftIcon class="h-6 w-6 ml-0 mr-1 text-100" /> Previous</div></template>
      <template #next><div class="flex">Next <ArrowRightIcon class="h-6 w-6 ml-1 mr-0 text-100" /></div></template>
      <template #finish><div class="flex">Finish <CheckIcon class="h-6 w-6 ml-1 mr-0 text-100" /></div></template>
    </FormStepsControls>
  </template>
</Vueform>
```

As you can see here, we're submitting the form with the default functionality using the Finish button. The request is sent to the API as expected and it returns a JSON object on success or on failure, in which case there are one or more error messages to display to the user. I have the [FormErrors](https://vueform.com/reference/form-errors) at the top of the template and can see the red box where the messages would go, but I could use some help with actually wiring that up.

## Screenshots

Form-level:
<img src="https://github.com/vueform/vueform/assets/1206/b8f950af-8d0e-4fe8-b73d-355a0490d74f" alt="Vueform - form errors" width="480" />

Field-level:
<img src="https://github.com/vueform/vueform/assets/1206/2a1be47e-c26f-42dd-a024-d4726dde5b26" alt="Vueform - field error messages" width="480" />

**Comment by adamberecz** [comment, marked as answer]
> First off, I'm really enjoying working with Vueform. There's a lot of nuanced functionality without feeling like it's over-engineered. Very cool 🙆

That's great to hear!

It seems like `form$.messageBag.append` is what you're looking for:
https://vueform.com/docs/validating-elements#custom-errors-and-messages

**Reply by delano** [reply]
Ah yeah, I'd missed that. Thank you. 

Is there a canonical place where that logic should live when the response comes back from the server? Is it `handleSubmit` or `@submit` or? Right now for this particular case, I use a `@click` and send the request with axios manually.

**Reply by adamberecz** [reply]
It's better to use [`@submit`](https://vueform.com/reference/vueform#event-submit) or [`:endpoint`](https://vueform.com/reference/vueform#option-endpoint) (as a function) to handle form submission manually.

The questions is if you really need manual handling. If you just want to send validated data to an endpoint once the form is [submitted](https://vueform.com/reference/button-element#option-submits), setting `:endpoint` as a string will do exactly that. Then you can take care of response handling by subscribing to [`@success`](https://vueform.com/reference/vueform#event-success) and [`@error`](https://vueform.com/reference/vueform#event-error), or unifiedly [`@response`](https://vueform.com/reference/vueform#event-success) event.
